### PR TITLE
Add feature flags to the config for Wiki and Issues

### LINKED ISSUE #3336: [feature] Ability to enable/disable wiki system and issue tracking for all repositories
- Gitea version (or commit ref): 1.3.2
- Git version: 2.9.3
- Operating system: Red Hat Enterprise Linux Server release 7.4
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x ] Not relevant
- Log gist:

## Description
It would be nice to have a configuration option that allows administrators to enable and disable wiki system and issue tracking of all repositories.

I could manage to disable issue tracking(not wiki system) of all the repositories by executing following SQL statements:
```
DELETE FROM repo_unit WHERE type=2;
UPDATE repo_unit SET config='{}';
```
P.S.: Similar request for Gogs: https://github.com/gogits/gogs/issues/3738

### PR DESCRIPTION
**Pull request by jpicht**
This pull request aims to close #3336, implementing ENABLE_WIKI and ENABLE_ISSUES config options (defaulting to true).

It would probably be good if somebody, who really knows how the Issue system works, had a close look over what I did and if there are some places that I missed (maybe pull-request related).

Currently there are two unrelated commits in this branch, that I needed to make the Makefile work for me. I can remove them, if they pose a problem.

### CODE REVIEW

**Comment by jpicht** [comment]
The next step for me would be (in another pull request) to try and make pull requests work with issues disabled.

**Comment by lunny** [comment]
I think just change `defaultRepoUnits` according configs and change the repository settings UI is OK. Or maybe team add/edit page.

**Comment by jpicht** [comment]
> I think just change `defaultRepoUnits` according configs and change the repository settings UI is OK. Or maybe team add/edit page.

I thought about that, but I was unsure if the config file is parsed early enough. I do not really know enough about the initialization sequence. When some code path uses `defaultRepoUnits` before the config is read, it could have weird consequences.

I can change the patch set to do exactly what you said, if that's the better solution.

**Comment by charlesmorin** [comment]
Do we have any ETA for this feature to be pulled into the master branch? Working in a large organization where Confluence and JIRA are used, such flag would be very helpful. Thanks!

File: Makefile [hunk 1]
```diff
@@ -22,7 +22,7 @@ BINDATA := modules/{options,public,templates}/bindata.go
 GOFILES := $(shell find . -name "*.go" -type f ! -path "./vendor/*" ! -path "*/bindata.go")
 GOFMT ?= gofmt -s
```
Comment (techknowlogick) [inline Makefile]: Please revert changes to Makefile, as these changes appear to be out of scope for this PR. If you'd still like to have these and they are indeed unrelated please open a separate PR just for them where they can be discussed/reviewed.

File: routers/api/v1/api.go [hunk 2]
```diff
@@ -407,7 +407,7 @@ func orgAssignment(args ...bool) macaron.Handler {
 }
 
 func mustEnableIssues(ctx *context.APIContext) {
-	if !ctx.Repo.CanRead(models.UnitTypeIssues) {
+	if !setting.Service.EnableIssues || !ctx.Repo.CanRead(models.UnitTypeIssues) {
```
Comment (lunny) [inline routers/api/v1/api.go]: You can check that on `CanRead` or more deep function but not here.

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/repo/issue_label_test.go [hunk 3]
```diff
@@ -119,6 +120,7 @@ func TestDeleteLabel(t *testing.T) {
 }
 
 func TestUpdateIssueLabel_Clear(t *testing.T) {
+	setting.Service.EnableIssues = true
```
Comment (lunny) [inline routers/repo/issue_label_test.go]: Since default value is true, this is unnecessary.

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/repo/issue_label_test.go (hunk 3 above)
Comment (jpicht) [reply routers/repo/issue_label_test.go]: Actually the tests fail, if this line is removed. (See https://drone.gitea.io/go-gitea/gitea/10658/1/10)

**Comment by jpicht** [comment]
Now the docker test fails because of unrelated issues. :(

Can that be re-run or do I need to push an empty commit to the branch?

```
go: strk.kbt.io/projects/go/libravatar@v0.0.0-20160628055650-5eed7bff870a: git fetch -f https://strk.kbt.io/git/go-libravatar.git refs/heads/*:refs/heads/* refs/tags/*:refs/tags/* in /go/pkg/mod/cache/vcs/20b35e8f598eb07a4d3d0fb9a878597498193f8a12440138d148579fe7ea90eb: exit status 128:
--
418 | fatal: unable to access 'https://strk.kbt.io/git/go-libravatar.git/': The requested URL returned error: 500
```

**Comment by lafriks** [comment]
Currently this will totally break options page to enable/disable pull requests if both issue and wiki are disabled, no?

IMHO this PR can not be merged if it leaves broken state that no PR can be used in Gitea

**Comment by jpicht** [comment]
> Currently this will totally break options page to enable/disable pull requests if both issue and wiki are disabled, no?
Actually Pull-Requests are working just fine. I don't know about a broken options page? Can you point me to it?

**Comment by jpicht** [comment]
> > Currently this will totally break options page to enable/disable pull requests if both issue and wiki are disabled, no?
> 
> Actually Pull-Requests are working just fine. I don't know about a broken options page? Can you point me to it?

Never mind ... I found it.

File: modules/setting/service.go [hunk 4]
```diff
@@ -83,6 +90,10 @@ func newService() {
 	Service.DefaultOrgVisibility = sec.Key("DEFAULT_ORG_VISIBILITY").In("public", structs.ExtractKeysFromMapString(structs.VisibilityModes))
 	Service.DefaultOrgVisibilityMode = structs.VisibilityModes[Service.DefaultOrgVisibility]
 
+	// feature flags
+	Service.EnableIssues = sec.Key("ENABLE_ISSUES").MustBool(true)
```
Comment (zeripath) [inline modules/setting/service.go]: I don't think these belong in Service. They would be better in Repository

**Review by zeripath — CHANGES_REQUESTED** [review]

**Comment by jpicht** [comment]
@zeripath I changed the config section to [repository]

**Review by zeripath — APPROVED** [review]

**Comment by jpicht** [comment]
@lunny Somehow the page still shows one of your change request to be open. It somehow got confused, when I rebased the branch on the current master. The requested change has been implemented.

File: models/repo.go [hunk 5]
```diff
@@ -1278,6 +1278,9 @@ func createRepository(e *xorm.Session, doer, u *User, repo *Repository) (err err
 	var units = make([]RepoUnit, 0, len(DefaultRepoUnits))
 	for _, tp := range DefaultRepoUnits {
 		if tp == UnitTypeIssues {
+			if !setting.Repository.EnableIssues {
```
Comment (lafriks) [inline models/repo.go:1281]: wouldn't it be possible to do these changes without changing how unit rights are saved to DB? so that if you later enable enable issues back you would not have to go through all repo teams to enable issues back

Comment (jpicht) [reply models/repo.go:1281]: I don't really think this is will happen.

Example: We have a company wiki, and an issue tracker. All we do need gitea for is for managing git repositories.
Enabling wiki and issue functions will - very probably - not happen in this use case.

I also don't see a different use case, where you would want to globally disable the Wiki / Issue functions, do you?

DefaultRepoUnits also does not contain UnitTypeIssues (as by @lunny's request https://github.com/go-gitea/gitea/pull/7032#issuecomment-495650639). So that even removing that line will not change the behavior.

I can provide a small bash+curl script to enable the wiki and/or issues via the API for inclusion in the documentation, so a user could easily work around this.

File: models/repo.go [hunk 6]
```diff
@@ -1287,6 +1290,10 @@ func createRepository(e *xorm.Session, doer, u *User, repo *Repository) (err err
 					EnableDependencies:               setting.Service.DefaultEnableDependencies,
 				},
 			})
+		} else if tp == UnitTypeWiki {
+			if !setting.Repository.EnableWiki {
+				continue
+			}
```
Comment (davidsvantesson) [inline models/repo.go]: Unit is never appended

**Review by davidsvantesson — COMMENTED** [review]
Personally I find it more useful with a setting for default repo units. Even if it is a common use-case to not use internal wiki/tracker, you might want to link to an external one.

This PR still leaves it open to enable issues/wiki from API, is that intentional?

**Comment by jpicht** [comment]
> Personally I find it more useful with a setting for default repo units. Even if it is a common use-case to not use internal wiki/tracker, you might want to link to an external one.

I can certainly add "DEFAULT_WIKI_ENABLED" and "DEFAULT_ISSUES_ENABLED" config options, but including this in this pull request would be be scope creep IMO

> This PR still leaves it open to enable issues/wiki from API, is that intentional?
No, that would not be intentional. I thought I disabled the API routes, but I will check again.

File: models/repo.go (hunk 5 above)
Comment (lafriks) [reply models/repo.go:1281]: What I mean is what if I already have existing repository and now I disable issues globally, units for existing repositories will be added so there is not really much point in skiping this for new repos as you will have to check anyway later not only if unit is enabled but also new setting. And if for whaterver reason I later want to enable it back it would be better that user would not have to run scripts to fix db to have working instance.

**Comment by davidsvantesson** [comment]
> No, that would not be intentional. I thought I disabled the API routes, but I will check again.

Sorry, it might be a miss of me.

**Comment by davidsvantesson** [comment]
> > This PR still leaves it open to enable issues/wiki from API, is that intentional?
> 
> No, that would not be intentional. I thought I disabled the API routes, but I will check again.

I think https://try.gitea.io/api/swagger#/repository/repoEdit is still available which allows to enable or disable wiki/issue tracker? And I don't think the route shall be disabled because it would disable any repository setting through API. Rather an http error response shall be returned if trying to enable issue/wiki (alternatively, just ignore the setting).

File: models/repo.go (hunk 5 above)
Comment (davidsvantesson) [reply models/repo.go:1281]: To me it seems this PR is doing two things at the same time
- Disabling the possibility to enable issues/wiki (including default value)
- Disabling any issue/wiki view, even if it was enabled.

Of course in the second case, it would be confusing to still have the settings for enabling it, but there is really no point of affecting the defaults (as @lafriks point out). If only the first case is wanted, there should not be necessary with any changes in routers, as the wiki/issues is anyhow disabled. For a fresh installation I don't think it matter, but the question is how it shall behave for an existing installation if issues/wikis are globally disabled. Either issues/wikis are disabled for existing repos even if they was enabled before the ini setting was changed. Other alternative is to only allow disable issues/wiki for a repo but not enable it. This would correspond to how FORCE_PRIVATE option is working.

**Comment by lunny** [comment]
@jpicht sorry for late to review. I will review this ASAP.

File: routers/api/v1/api.go [hunk 7]
```diff
@@ -454,7 +454,7 @@ func mustAllowPulls(ctx *context.APIContext) {
 }
 
 func mustEnableIssuesOrPulls(ctx *context.APIContext) {
-	if !ctx.Repo.CanRead(models.UnitTypeIssues) &&
+	if (!setting.Repository.EnableIssues || !ctx.Repo.CanRead(models.UnitTypeIssues)) &&
```
Comment (lunny) [inline routers/api/v1/api.go:457]: So even the repository created some issues, this will still hide them when `setting.Repository.EnableIssues` changed to true?

**Comment by 6543** [comment]
@davidsvantesson are you still working on this?

**Comment by 6543** [comment]
@jpicht are you still working on this?

**Comment by jpicht** [comment]
We are running this in production, so I would be happy to get it merged. I've ported it to the latest rc one or two weeks back.

The problem with merging though, seems to mostly be to define what is supposed to happen in the corner cases, that (IMO) would probably never happen: re-enabling issues/wiki after disabling them.

I do only see the use case that my company has: we already have a wiki and an issue tracker, we do not want to confuse people with a second one, that they are supposed to ignore. If there's another use case I'd be happy to accommodate that in my patch.

I'll push an updated version in the next few days.

**Comment by guillep2k** [comment]
@jpicht In the meantime, you could cheat the feature by using a custom template that _hides_ Wikis and Issues from the menu.

**Comment by davidsvantesson** [comment]
I don't think the corner cases is a big problem. My main comment on this PR is that I think it solves a  too specific use case. Shouldn't we have a system where any repo unit can be globally disabled? (Maybe a bit connected to the plugin discussions). So the app.ini could be something like
`
GLOBAL_UNIT_DISABLE = issues, external_tracker, wiki, external_wiki, pull_requests, releases
`

I don't mean disabling of all units need to be implemented in the same PR, just that it would be good to think about the structure. 

Sorry if I am being discouraging, I really see the use of being able to globally disable certain units. In my organization we don't use the internal issue and wiki, so it would be good to be able to disable those (we have a script to turn them off at creation). The problem is that this PR also switches off other units with the same flag options (external tracker and wiki), which makes it much less useful.

**Comment by davidsvantesson** [comment]
I think this pull request can be closed as #8788 solves the same thing.

@jpicht Thanks for your work! It is valuable even if a slightly different solution was merged.
